### ISSUE TITLE
Axis scaling doesn't consider outside labels in series

### ISSUE THREAD

**Issue by crypto-rsa** [oxyplot/oxyplot#1776]
### Steps to reproduce

1. Open the BarSeries - BaseValue (labels) example

### Expected behaviour
The labels of all series should be visible, even when they are placed outside of the bar items.

### Actual behaviour
Outside labels are not visible with the default axis scaling.
![obrazek](https://user-images.githubusercontent.com/11479798/128511083-3f34673b-1b20-43ab-a85e-286e749f0f05.png)

**Comment by VisualMelon** [comment]
This is impossible in data space, but using the new screen-space margins, it might be that we can compute automatic screen-space margins from series etc. (i.e. bar series can ask for 50px of screen space, line series can ask for none, etc.

**Comment by crypto-rsa** [comment]
I am not familiar with the code but I suppose that at rendering time, we know what needs to be rendered _outside_ the series so we can subtract that from the available space and do the axis scaling after that, can't we?

**Comment by VisualMelon** [comment]
If label placement is consistent enough, then that could be possible. It's not as simple as subtracting what is outside, because what is outside depends on the axes (e.g. at one scale, a short label may escape the plot area, at another, a long label may escape the plot area; it's a system of linear equations we need to solve (can simulate data-bounds as zero-width so they don't need to be special cases)). OxyPlot doesn't keep track of rendered elements like labels at the moment, so it would require some work:
 - need to distinguish elements that may/may not be clipped
 - need to record this information (including: axis pair, data-space location, (v/h) alignment, screen-space dimensions and offsets, ?? anything else?)
   - how does this information get to the axis?
     - could add to `IAxis/AxisBase`, and axes keep track of them individually
     - could accumulate everything in `PlotModel`, and the axes query for it
       - is there any advantage to PlotModel knowing about them for e.g. interaction purposes?
         - (at the moment, non-annotation-based labels don't have any consistency, so we can't e.g. make them interactive)
         - 
 - perform the rescaling you describe
   - not immediately obvious how to do this efficiently
     - this would have to be run inside the margin loop, so it needs to be efficient
   - should be easy enough to make it work with H/V axis, anything else may be problematic (e.g. orthogonal but not screen-aligned, or polar) (can just not try if necessary, not a priority)
   - Would be in `UpdateActualMaxMin`; can't be performed for min/max independently
     - easiest to just bolt it onto the end as a another stage which only runs if _nothing_ is explicitly set (minimum/viewminimum, etc.)
     - Must decide (or make configurable) how this would interact with existing axis margins

This sort of thing is useful for flat plots that are rendered once, but can be a sort of irritation if the same plot is rendered with different data (min/max margins are better for that sort of thing). 

Possible benefits:
 - Easier to generate non-misleading/ugly plots automatically
 - Recording label information means they can participate in interaction more easily

Possible additional complexities:
 - A label arrangement system (to prevent overlap) would fight with this

Where might such labels be applicable:
 - Bar series, per this issue
 - Labels on scatters/lines (equally, they might want to be arranged onto the plot area rather than expand it, which is better when you have known min/max)
 - Custom text annotations (configurable)
 - Not applicable to any screen-space informed labels (e.g. path annotations which clamp to the edge of the screen)

<hr />

Mysterious notes:
 - Would translate to interaction space immediately, then everything should be the same

**Comment by crypto-rsa** [comment]
Whoa, sounds way more complicated than I expected!
Is there some workaround to make the axis scale a bit lower from the outside, so that the labels are not truncated (albeit in a very rough way)?

**Comment by crypto-rsa** [comment]
Also, another issue possibly related to this (because it would affect the axis scaling) is rendering the plot title/subtitle. Currently if the plot is not wide enough, the title is rendered at multiple lines but only the first line is visible and the rest is obscured by the plot itself:
![obrazek](https://user-images.githubusercontent.com/11479798/128674952-41d3d0eb-d54d-4eb3-ac3f-aafa87c86479.png)

**Comment by VisualMelon** [comment]
I don't think it would be that hard to implement, actually, assuming there is a simple and efficient way to perform the rescaling, but there are a lot of decisions to be made, which makes means it could take a lot of time and discussion and testing to get right.

You can increase the data-space padding (e.g. `Axis.MaximumPadding`, default is 2% I think); these are complemented in 2.1 with screen space variants (e.g. `Axis.MaximumDataMargin`, default is 0).

The title weirdness is probably a different issue; there are some issues with multi-line text on some platforms. Feel free to open a separate issue for that with some example code and I'll take a look when I have time.

**Comment by crypto-rsa** [comment]
#1778 logged.

**Comment by crypto-rsa** [comment]
The issue also applies to outside labels in pie series.
